subsys/mgmt/mcumgr/transport/src/smp_lorawan.c

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2024, Jamie McCrae
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/lorawan/lorawan.h>
+#include <zephyr/mgmt/mcumgr/smp/smp.h>
+#include <zephyr/mgmt/mcumgr/transport/smp.h>
+#include <zephyr/mgmt/mcumgr/mgmt/handlers.h>
+
+#include <mgmt/mcumgr/transport/smp_internal.h>
+#include <mgmt/mcumgr/transport/smp_reassembly.h>
+
+LOG_MODULE_REGISTER(smp_lorawan, CONFIG_MCUMGR_TRANSPORT_LORAWAN_LOG_LEVEL);
+
+static void smp_lorawan_downlink(uint8_t port, bool data_pending, int16_t rssi, int8_t snr,
+				 uint8_t len, const uint8_t *hex_data);
+
+static int smp_lorawan_uplink(struct net_buf *nb);
+
+static uint16_t smp_lorawan_get_mtu(const struct net_buf *nb);
+
+static struct lorawan_downlink_cb lorawan_smp_downlink_cb = {
+	.port = CONFIG_MCUMGR_TRANSPORT_LORAWAN_FRAME_PORT,
+	.cb = smp_lorawan_downlink
+};
+
+struct smp_transport smp_lorawan_transport = {
+	.functions.output = smp_lorawan_uplink,
+	.functions.get_mtu = smp_lorawan_get_mtu,
+};
+
+#ifdef CONFIG_SMP_CLIENT
+struct smp_client_transport_entry smp_lorawan_client_transport = {
+	.smpt = &smp_lorawan_transport;
+	.smpt_type = SMP_LORAWAN_TRANSPORT;
+};
+#endif
+
+#ifdef CONFIG_MCUMGR_TRANSPORT_LORAWAN_POLL_FOR_DATA
+static struct k_thread smp_lorawan_thread;
+K_KERNEL_STACK_MEMBER(smp_lorawan_stack, CONFIG_MCUMGR_TRANSPORT_LORAWAN_POLL_FOR_DATA_STACK_SIZE);
+K_FIFO_DEFINE(smp_lorawan_fifo);
+
+struct smp_lorawan_uplink_message_t {
+	void *fifo_reserved;
+	struct net_buf *nb;
+	struct k_sem my_sem;
+};
+
+static struct smp_lorawan_uplink_message_t empty_message = {
+	.nb = NULL,
+};
+
+static void smp_lorawan_uplink_thread(void *p1, void *p2, void *p3)
+{
+	struct smp_lorawan_uplink_message_t *msg;
+
+	while (1) {
+		msg = k_fifo_get(&smp_lorawan_fifo, K_FOREVER);
+		uint16_t size = 0;
+		uint16_t pos = 0;
+
+		if (msg->nb != NULL) {
+			size = msg->nb->len;
+		}
+
+		while (pos < size || size == 0) {
+			uint8_t *data = NULL;
+			uint8_t data_size;
+			uint8_t temp;
+			uint8_t tries = CONFIG_MCUMGR_TRANSPORT_LORAWAN_POLL_FOR_DATA_RETRIES;
+
+			lorawan_get_payload_sizes(&data_size, &temp);
+
+			if (data_size > size) {
+				data_size = size;
+			}
+
+			if (size > 0) {
+				data = net_buf_pull_mem(msg->nb, data_size);
+			}
+
+			while (tries > 0) {
+				int rc;
+
+				rc = lorawan_send(CONFIG_MCUMGR_TRANSPORT_LORAWAN_FRAME_PORT,
+						  data, data_size,
```

Comment:
If data_size > size, then `data = NULL` then the `lorawan_send` will return -EINVAL, this will be counted as -- to 'tries', that is correct yes?